File: Assign1/task2.scala
object task2_1{
        def main(args:Array[String]):Unit={
        if(args(3)=="spark"){
        val conf=new SparkConf().setAppName("spark-test").setMaster("local[*]")
        val sc=new SparkContext(conf)
        val textFile=sc.textFile(args(0))
        val review_json=textFile.map(row=>parse(row))
        val review_row=review_json.map(x=>((x \ "business_id").values.toString,(x \ "stars").values.toString))

        val business_data=sc.textFile(args(1))
        val business_j=business_data.map(x=>parse(x))

        val business_json=business_j.filter(x=>(x \ "categories")!=JNull).flatMap(x=>for(n<-(x \\ "categories").values.toString.split(','))yield{((x \\ "business_id").values.toString,n.trim())})
        val join1=review_row.join(business_json)
        val reduce=join1.map(x=>(x._2._2,(x._2._1,1))).reduceByKey((x,y)=>((x._1.toDouble+y._1.toDouble).toString,y._2.toInt+x._2.toInt)).map(x=>(x._1,(x._2._1.toDouble/x._2._2).toDouble)).sortBy(x=>(-x._2,x._1)).map(x=>List(x._1,x._2))

        val ans="result"->reduce.take(args(4).toInt).toList
        val file=new File(args(2))
        val bw=new BufferedWriter(new FileWriter(file))

        bw.write(Json(DefaultFormats).write(ans).replace(":",": ").replace(",",", "))
        bw.close()
        sc.stop()


        }else{

        //NOSPARK IMPLEMENTATION
        var business_map=Map[String,List[String]]()
        val textFile=Source.fromFile(args(1))
        for(i<-textFile.getLines){
        val line=parse(i)
        if((line \ "categories")!=JNull&&(line \ "categories")!=""){
        for(category<-(line \ "categories").values.toString.split(", ")){
        val business_id=(line \ "business_id").values.toString
        if(!business_map.contains(category)){
        business_map+=category->List(business_id)
        }else{
        val main=business_map(category):::List(business_id)
        business_map+=category->main
        }

        }
        }
        }
        var review_map=Map[String,List[Any]]()
        val review_file=Source.fromFile(args(0))
        for(i<-review_file.getLines){
        val line=parse(i)
        val business_id=(line \ "business_id").values.toString
        val stars=(line \ "stars").values.toString.toDouble
        if(!review_map.contains(business_id)){
        review_map+=business_id->List(stars,1.0)
        }else{
        val main=review_map(business_id)
        val newValue=List(main(0).toString.toDouble+stars,main(1).toString.toDouble+1.0)
        review_map+=business_id->newValue
        }

        }
        var ans_map=List[List[Any]]()
        business_map.keys.foreach{i=>
        var sum=0.0
        var count=0.0
        for(ids<-business_map(i)){
        if(review_map.contains(ids)){
        sum=sum+review_map(ids)(0).toString.toDouble
        count=count+review_map(ids)(1).toString.toDouble
        }
        }
        if(count!=0.0){
        ans_map=ans_map:::List(List(i,(sum/count)))
        }
        }
        val ans=ans_map.sortBy(x=>(x(1).toString.toDouble,x(0).toString))(Ordering.Tuple2(Ordering.Double.reverse,Ordering.String)).slice(0,args(4).toInt)
        val file=new File(args(2))
        val bw=new BufferedWriter(new FileWriter(file))

        bw.write(Json(DefaultFormats).write(ans).replace(":",": ").replace(",",", "))
        bw.close()

        }

        }
        }

File: Assign1/task2.py
from pyspark import SparkContext
import json
from collections import defaultdict 
import sys 

top_n = int(sys.argv[5])

if sys.argv[4] == "spark": 
    sc1 = SparkContext.getOrCreate()
    review_data = sc1.textFile(sys.argv[1]).map(lambda k: json.loads(k))
    business_data = sc1.textFile(sys.argv[2]).map(lambda k: json.loads(k))
    business_json = business_data.filter(lambda x : x['categories'] is not None).flatMap(lambda x: [(x['business_id'], y) for y in x['categories'].split(", ")]) 
    review_json = review_data.map(lambda x: (x['business_id'],x['stars']))

    join = review_json.leftOuterJoin(business_json)
    reduce = join.map(lambda x :(x[1][1],(x[1][0],1))).reduceByKey(lambda x,y : (x[0]+y[0], x[1]+y[1],2)).map(lambda x : (x[0],float(x[1][0]/x[1][1]))).sortBy(lambda x : (-x[1],x[0]))

    result = {
        "result" : reduce.take(top_n)
    }
    f = open(sys.argv[3],"w")
    json.dump(result, f)
    f.close()
    sc1.stop()

else : 
    print(sys.argv[4])
    f = open(sys.argv[1])
    lines = f.readlines()
    review_list=dict()
    for line in lines: 
        review_data = json.loads(line)
        if(review_list.get(review_data['business_id']) == None) :
            review_list[review_data['business_id']]= [review_data['stars'],1]
        else : 
            review_list[review_data['business_id']] = [review_list[review_data['business_id']][0]+ review_data['stars'], 1+review_list[review_data['business_id']][1]]
        

    f = open(sys.argv[2])
    lines = f.readlines()
    business_list=[]
    category = defaultdict(list)
    for line in lines:
        business_data = json.loads(line)
        if(business_data['categories']!= None) :
            categories = business_data['categories'].split(", ")
            for i in categories: 
                category[i].append(business_data['business_id'])
    avg_dict = defaultdict()
    for category,business_ids in category.items() :
        sum = 0 
        count = 0
        for id in business_ids : 
            if(review_list.get(id) !=None) : 
                    sum = sum + review_list[id][0]
                    count = count + review_list[id][1]  
        if(count != 0):
            avg_dict[category] = sum/count

    ans_dict = sorted(avg_dict.items(), key = lambda x : (-x[1],x[0]))
    result = {
        "result" : ans_dict[:top_n]
    }
    f = open(sys.argv[3],"w")
    json.dump(result, f)
    f.close()



